// sass-embedded/dist/tool/get-embedded-compiler.js
"use strict";
// Copyright 2022 Google Inc. Use of this source code is governed by an
// MIT-style license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.
Object.defineProperty(exports, "__esModule", { value: true });
exports.getEmbeddedCompiler = void 0;
const fs_1 = require("fs");
const p = require("path");
const yaml = require("yaml");
const shell = require("shelljs");
const utils = require("./utils");
/**
 * Downlaods and builds the Embedded Dart Sass compiler.
 *
 * Can check out and build the source from a Git `ref` or build from the source
 * at `path`. By default, checks out the latest revision from GitHub.
 */
async function getEmbeddedCompiler(outPath, options) {
    var _a;
    const repo = 'dart-sass-embedded';
    let source;
    if (!options || 'ref' in options) {
        utils.fetchRepo({
            repo,
            outPath: 'build',
            ref: (_a = options === null || options === void 0 ? void 0 : options.ref) !== null && _a !== void 0 ? _a : 'main',
        });
        source = p.join('build', repo);
        await maybeOverrideSassDependency(source);
    }
    else {
        source = options.path;
    }
    buildDartSassEmbedded(source);
    await utils.link(p.join(source, 'build'), p.join(outPath, repo));
}
exports.getEmbeddedCompiler = getEmbeddedCompiler;
/**
 * Overrides Embedded Dart Sass compiler's dependency on Dart Sass to use the
 * latest version of Dart Sass from the `main` branch.
 *
 * This allows us to avoid needing to commit a dependency override to the
 * embedded compiler when it doesn't actually require any local changes.
 */
async function maybeOverrideSassDependency(repo) {
    const pubspecPath = p.join(repo, 'pubspec.yaml');
    const pubspec = yaml.parse(await fs_1.promises.readFile(pubspecPath, { encoding: 'utf-8' }));
    console.log(`Overriding ${repo} to load Dart Sass from HEAD.`);
    pubspec['dependency_overrides'] = {
        ...pubspec['dependency_overrides'],
        sass: { git: 'https://github.com/sass/dart-sass.git' },
    };
    await fs_1.promises.writeFile(pubspecPath, yaml.stringify(pubspec), { encoding: 'utf-8' });
}
// Builds the Embedded Dart Sass executable from the source at `repoPath`.
function buildDartSassEmbedded(repoPath) {
    console.log('Downloading dart-sass-embedded dependencies.');
    shell.exec('dart pub upgrade', {
        cwd: repoPath,
        silent: true,
    });
    console.log('Building dart-sass-embedded executable.');
    shell.exec('dart run grinder protobuf pkg-standalone-dev', {
        cwd: repoPath,
        silent: true,
    });
}
